Add Comparison, Equivalence, Predicate

garyb opened the pull request:
I'll add some comments before merging, but just wanted some feedback first.

Do we think these all belong together? `Predicate`, `Equivalence` etc. could actually live in their own packages potentially.

garyb committed:
Fix bower.json

Add flipped contramap

Update grunt plugin

Add Comparison, Equivalence, Predicate

Relax Ord constraint to Eq

Move under Data.Functor

Better contramap implementations

Add Compose and friends

Update docs

Update docs

Remove compose

garyb commented:
/cc @paf31

paf31 commented:
:+1: I like it.

joneshf commented:
Very nice!

joneshf commented on src/Data/Equivalence.purs:
```diff
@@ -0,0 +1,25 @@
+instance monoidEquivalence :: Monoid (Equivalence a) where
+  mempty = Equivalence (\_ _ -> true)
+
+defaultEquivalence :: forall a. (Ord a) => Equivalence a
```
You only need `Eq a` here right?

garyb replied:
Oops, yep. Thanks.

garyb commented:
Added some docs now, and moved the module to `Data.Functor.Contravariant` - I'll be making a repo for invariant functors soon too, so figured it'd make sense to group them both under `Data.Functor`.

Thoughts? And how do the docs look? @paf31 @joneshf

paf31 commented:
Looks great. Does it make sense to move `Compose` out into its own library? It has other instances which could be useful, like `Functor` and `Applicative`.

garyb commented:
Sure, I can do that.

I felt similarly about `Comparison` and `Equivalence`, but figure I'll leave those for now, and perhaps when we dismantle the `Prelude` they can live with `Ord` and `Eq` or something.

joneshf commented:
I thought `Compose` was already a thing! :) I agree with moving it to its own library. Or maybe just put it in `purescript-control`? 

I don't really understand the namespace change, but then I never have understood the namespaces.

joneshf commented on src/Data/Functor/Contravariant/Compose.purs:
```diff
@@ -0,0 +1,36 @@
+  (<$>) f (Compose x) = Compose ((f >$<) >$< x)
+
+-- | Composition of a covariant and contravariant functor.
+newtype ComposeFC f g a = ComposeFC (f (g a))
```
Is there a reason for both `ComposeFC` and `ComposeCF`? Overlapping instances or some such?

garyb replied:
Good point, having both versions might be unnecessary. Will give it a try after moving this out.

paf31 commented:
Sorry, I didn't think about the possible overlapping instances issue. I guess things are almost never going to be both covariant and contravariant, but it is a possible issue. Even if they are, the two instances should be identical, right? Maybe we just have to stick the two instances in different modules.
